### PR TITLE
Introduced parameter "base_dataset_id" for writing multi-level datasets

### PR DESCRIPTION
**Pull request by forman**
Introduced parameter "base_dataset_id" for writing multi-level datasets.

This is a first step to address #617. In a next step, `xcube level` CLI tool should use the data store framework.

Checklist:

* [x] Add unit tests and/or doctests in docstrings
* [x] Add docstrings and API docs for any new/modified user-facing classes and functions
* [x] Changes documented in `CHANGES.md`
* [x] AppVeyor CI passes
* [x] Test coverage remains or increases (target 100%)

Remember to close associated issues after merge!

### CODE REVIEW

File: xcube/core/store/fs/impl/dataset.py [hunk 1]
```diff
@@ -160,9 +162,12 @@ def get_open_data_params_schema(self, data_id: str = None) \
             ZARR_OPEN_DATA_PARAMS_SCHEMA
         )
 
-    def open_data(self, data_id: str, **open_params) -> xr.Dataset:
+    def open_data(self,
+                  data_id: str,
+                  root: Optional[str] = None,
```
Comment (TonioF) [inline xcube/core/store/fs/impl/dataset.py:167]: I think you should remove this here. It is later superseded by the root from load_fs, however, that one will always be None as root is not in open_params.

File: xcube/core/store/fs/impl/dataset.py [hunk 2]
```diff
@@ -212,8 +217,9 @@ def write_data(self,
 
     def delete_data(self,
                     data_id: str,
+                    root: Optional[str] = None,
```
Comment (TonioF) [inline xcube/core/store/fs/impl/dataset.py:220]: Same here

**Review by TonioF — CHANGES_REQUESTED** [review]
I still have the error with the inconsistent number of detail levels. Can you please set the chunking explicitly?

File: xcube/core/store/fs/impl/mldataset.py [hunk 3]
```diff
@@ -172,13 +213,13 @@ def _get_levels(cls,
                     data_id: str) -> List[int]:
         levels = []
         for dir_name in (os.path.basename(dir_path['name'])
-                         for dir_path in fs.listdir(data_id, detail=True)
-                         if dir_path['type'] == 'directory'):
+                         for dir_path in fs.listdir(data_id, detail=True)):
+            print("_get_levels:", dir_name)
```
[the remark this answers is not in the document]
Comment (forman) [reply xcube/core/store/fs/impl/mldataset.py:217]: Overlooked it.

File: xcube/core/store/fs/impl/dataset.py (hunk 1 above)
Comment (forman) [reply xcube/core/store/fs/impl/dataset.py:167]: Agreed!

File: xcube/core/store/fs/impl/mldataset.py [hunk 4]
```diff
@@ -91,55 +101,86 @@ def write_data(self,
...
-                level_dataset = xr.open_zarr(zarr_store,
-                                             consolidated=consolidated)
-                ml_dataset.set_dataset(index, level_dataset)
+            if base_dataset_id and index == 0:
+                # Write file "0.link" instead of copying
+                # level-0 dataset to "0.zarr"
+                link_path = f'{data_id}/{index}.link'
+                fs.mkdirs(data_id, exist_ok=replace)
+                with fs.open(link_path, mode='w') as fp:
+                    fp.write(base_dataset_id)
```
[the remark this answers is not in the document]
Comment (forman) [reply xcube/core/store/fs/impl/mldataset.py:110]: We could later add a parameter `check: bool = False` should that become important. The current implementation has the advantage that the dataset at the given link must not _yet_ exist.

**Review by TonioF — APPROVED** [review]
Okay, it's all working now.

**Comment by forman** [comment]
Merging although conda repo for Linux-64 seems broken. On Mac OS, AppVeyor is green.

**Pull request merged**